jpc/src/tag_tree.rs

```diff
@@ -0,0 +1,193 @@
+/// Level 0 is the lowest level of the tag tree; it contains the top node.
+pub struct TagTree {}
+
+/// A decoder from tag tree bits to numbers in the 2d-array.
+///
+/// TagTreeDecoder takes in bits and returns values from the represented 2d-array. Only positive
+/// integers will ever be produced.
+#[derive(Debug)]
+pub struct TagTreeDecoder {
+    max_depth: usize,
+    cur_depth: usize,
+    cur_value: u8,
+    /// levels is a Vec containing data (width, items) for each level of the tree.
+    levels: Vec<(usize, Vec<u8>)>, // Todo, what should be here?
+}
+
+impl TagTreeDecoder {
+    pub fn new(width: usize, height: usize) -> Self {
+        let mut mw = width;
+        let mut mh = height;
+        let mut max_depth = 0;
+        let mut levels = Vec::new();
+        // Determine max depth by dividing out groups of 4
+        while mw > 1 || mh > 1 {
+            let w = mw.max(1);
+            let size: usize = w * mh.max(1);
+            levels.push((w, Vec::with_capacity(size)));
+            println!("added vec of size {size}");
+            max_depth += 1;
+            mw = mw.div_ceil(2);
+            mh = mh.div_ceil(2);
+        }
+        levels.push((1, Vec::with_capacity(1)));
+        levels.reverse(); // reverse in place so level 0 is at index 0
+        println!("Need a depth of {max_depth} to represent tag tree");
+
+        assert_eq!(max_depth + 1, levels.len());
+
+        Self {
+            max_depth,
+            cur_depth: 0,
+            cur_value: 0,
+            levels,
+        }
+    }
+
+    fn cur_offset(&self) -> usize {
+        match self.levels.get(self.cur_depth) {
+            None => panic!("Not deep enough to know current location."),
+            Some((cw, clvl)) => clvl.len(),
+        }
+    }
+
+    // TODO fix push_bit type signature for better return type
+    pub fn push_bit(&mut self, b: u8) -> Option<u8> {
+        if b == 0 {
+            self.cur_value += 1;
+            return None;
+        }
+        assert_eq!(1, b);
+        // b == 1, record value at current position, prep return value, go to next position
+        let (_, lvl) = &mut self.levels[self.cur_depth];
+        lvl.push(self.cur_value);
+        // maintain cur_value and cur_depth to point at next position to fill
+        if self.cur_depth < self.max_depth {
+            // deeper!
+            self.cur_depth += 1;
+            info!(
+                "Recorded {} now at depth {}",
+                self.cur_value, self.cur_depth
+            );
+            return None;
+        }
+        if self.max_depth == 0 {
+            // handle single value tag tree... todo might be nicer some where else
+            return Some(self.cur_value);
+        }
+
+        // record out and walk up tree to maintain invariants
+        let out = self.cur_value;
+
+        // Find next place that needs a value and parent that provides cur_value
+        loop {
```

Comment:
I'm not sure about this logic, although I agree we need to traverse upwards I would expect this to only happen once we've filled in the nodes for that level.